### PR TITLE
Add windows/darwin support to npm_packages

### PR DESCRIPTION
**Pull request by terracatta**
## What Does This PR Do?

This PR does three things:

1. improves the `npm_packages` table so that it runs on every compatible osquery platform (including Windows and macOS). 
2. Adds the `homepage` key from the `package.json` file to the schema as a new column.
3. Improves the default directory search list based on my independent research finding the default paths of node modules on Mac, windows, and linux devices.

### Default run on macOS
```
osquery> select * from npm_packages;
       name = npm
    version = 8.5.5
description = a package manager for JavaScript
     author = Isaac Z. Schlueter <i@izs.me> (http://blog.izs.me)
    license = Artistic-2.0
   homepage = https://docs.npmjs.com/
       path = /opt/homebrew/lib/node_modules/npm/package.json
  directory = /opt/homebrew/lib/
```

###  Run on macOS with custom directory

```
osquery> select name, version, description from npm_packages where directory = "Users/jmeller/code/kolide/k2/" and name LIKE "c%" limit 3;
+----------------------+---------+-----------------------------------------------+
| name                 | version | description                                   |
+----------------------+---------+-----------------------------------------------+
| canvas-confetti      | 0.4.2   | performant confetti animation in the browser  |
| chart.js             | 2.9.4   | Simple HTML5 charts using the canvas element. |
| chartjs-color-string | 0.6.0   | Parser and generator for CSS color strings    |
+----------------------+---------+-----------------------------------------------+
```

### Default Run on Windows 10

```
osquery> SELECT * FROM npm_packages;
       name = tldr
    version = 3.3.7
description = Simplified and community-driven man pages
     author = Romain Prieto
    license = MIT
   homepage = http://tldr-pages.github.io
       path = C:\Users\jason\AppData\Roaming\npm\node_modules\tldr\package.json
  directory = C:\Users\jason\AppData\Roaming\npm\

       name = yarn
    version = 1.22.18
description = 📦🐈 Fast, reliable, and secure dependency management.
     author =
    license = BSD-2-Clause
   homepage =
       path = C:\Users\jason\AppData\Roaming\npm\node_modules\yarn\package.json
  directory = C:\Users\jason\AppData\Roaming\npm\
```

## Why Is This Useful?
There have been [a number of high-profile supply-chain attacks](https://www.mandiant.com/resources/supply-chain-node-js) involving npm_packages. While the current table can help locate compromised packages in production environments, it's also useful to scan for them on a developer's device directly. Many NPM packages can be installed globally and are available in the the user's $PATH. 

By expanding to more operating systems, we can drastically improve detection of these nefarious packages.

## Implementation
This PR mostly apes the cross-platform implementation of the `python_packages` table and combines in all the previous work done on this table under linux.

### CODE REVIEW

File: osquery/tables/system/npm_packages.cpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+#ifdef WIN32
+#include "windows/registry.h"
+#endif
+
+namespace fs = boost::filesystem;
+
+namespace osquery {
+namespace tables {
+
+const std::set<std::string> kNodeModulesPath = {
```
Comment (iko1) [inline osquery/tables/system/npm_packages.cpp:32]: Consider initializing this std::set differently by checking which OS the code is running.

File: osquery/tables/system/npm_packages.cpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+      }
+    }
+  }
+}
+
+void genNodeSiteDirectories(const std::string& site,
+                            QueryData& results,
+                            Logger& logger) {
+  std::vector<std::string> manifest_paths;
+  resolveFilePattern(site + "/node_modules/%/package.json", manifest_paths);
```
Comment (iko1) [inline osquery/tables/system/npm_packages.cpp:105]: consider initializing 2nd path as a boost::filesystem::path and then join both paths like that:
`boost::filesystem::path p("/node_modules/%/package.json")
resolveFilePattern(site / p, manifest_paths);`

File: osquery/tables/system/npm_packages.cpp (hunk 1 above)
Comment (directionless) [reply osquery/tables/system/npm_packages.cpp:32]: For what it's worth, I think it's a reasonable pattern to initialize them all this way. We do it pretty commonly.

Comment (terracatta) [reply osquery/tables/system/npm_packages.cpp:32]: @iko1 went with a simple `IFDEF` here for windows, I hope that is what you had in mind! Seph is right about POSIX, they would nearly be the exact same list (you can actually use homebrew on linux too if you like). So I think it's ok to group them together.

File: osquery/tables/system/npm_packages.cpp (hunk 2 above)
Comment (terracatta) [reply osquery/tables/system/npm_packages.cpp]: Good suggestion. Added!

File: osquery/tables/system/npm_packages.cpp [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+  if (isPlatform(PlatformType::TYPE_WINDOWS)) {
+    // Enumerate any system installed npm packages
+    auto installPathKey = "HKEY_LOCAL_MACHINE\\" + kWinNodeInstallKey;
+    genWinNodePackages(installPathKey, results, logger);
+
+    // Enumerate any user installed npm packages
+    installPathKey = "HKEY_USERS\\%\\" + kWinNodeInstallKey;
+    genWinNodePackages(installPathKey, results, logger);
+  }
```
[the remark this answers is not in the document]
Comment (terracatta) [reply osquery/tables/system/npm_packages.cpp]: Yep, you are 100% correct. I've added this to the correct branch in the parent `if` statement.

**Review by sharvilshah — COMMENTED** [review]
Thanks @terracatta! This is super useful -- I think there is only one small merge conflict to be resolved

**Comment by terracatta** [comment]
@sharvilshah I fixed the conflict, I'm not sure why that one test failed it looks like it was just a random timeout?

**Comment by Smjert** [comment]
> @sharvilshah I fixed the conflict, I'm not sure why that one test failed it looks like it was just a random timeout?

I've kicked a new build and yeah, it's a known flaky test.

**Comment by directionless** [comment]
GitHub status says mac actions are being slooooowwwww

**Review by sharvilshah — APPROVED** [review]

**Pull request merged**